File: fibonacci/fib.py
"""
Heru Handika
4 November 2020
Fibonacci using memoization
"""

from functools import lru_cache

@lru_cache(maxsize=None)
def fib(n: int) -> int:
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

def fib_iter(n: int) -> int:
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

def main():
    n: int = input("Enter number:")
    print(fib(int(n)))
    print(fib_iter(40))

if __name__ == "__main__":
    main()

File: fibonacci/fib.rs
/*
Heru Handika
4 November 2020
Memoization fibonacci
*/

fn main() {
    let n: i64 = 40;

    println!("{}", fib_iter(n));

}

fn fib_iter(n: i64) -> i64 {
    if n < 2 {
        n
    } else {
        fib_iter(n - 2) + fib_iter(n - 1)
    }
}